File: packages/b2c-tooling-sdk/src/clients/metrics.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Metrics API client for B2C Commerce (Observability).
 *
 * > **CLOSED BETA:** The Metrics API (`observability/metrics/v1`) is a closed
 * > beta capability. Availability, request/response shapes, and OAuth scopes may
 * > change without notice, and the API must be enabled for your organization.
 *
 * Provides a fully typed client for the SCAPI Observability Metrics API using
 * openapi-fetch with OAuth authentication middleware. The API exposes
 * time-series operational metrics (overall, sales, eCDN, third-party services,
 * SCAPI, SCAPI hooks, MRT, controllers, and OCAPI) for an organization.
 *
 * @module clients/metrics
 */
import createClient, {type Client} from 'openapi-fetch';
import type {AuthStrategy} from '../auth/types.js';
import {OAuthStrategy} from '../auth/oauth.js';
import {JwtOAuthStrategy} from '../auth/oauth-jwt.js';
import type {paths, components} from './metrics.generated.js';
import {createAuthMiddleware, createLoggingMiddleware} from './middleware.js';
import {globalMiddlewareRegistry, type MiddlewareRegistry} from './middleware-registry.js';
import {buildTenantScope} from './custom-apis.js';

/**
 * Re-export generated types for external use.
 */
export type {paths, components};

/**
 * The typed Metrics API client for retrieving observability metrics.
 *
 * **CLOSED BETA** — see the module description.
 *
 * ## Common Endpoints
 *
 * | Method | Path | Description |
 * |--------|------|-------------|
 * | GET | `/organizations/{organizationId}/metrics/overall` | Overall application metrics |
 * | GET | `/organizations/{organizationId}/metrics/sales` | Sales metrics |
 * | GET | `/organizations/{organizationId}/metrics/scapi` | SCAPI request metrics |
 *
 * @example
 * ```typescript
 * import { createMetricsClient, toOrganizationId } from '@salesforce/b2c-tooling-sdk/clients';
 * import { OAuthStrategy } from '@salesforce/b2c-tooling-sdk/auth';
 *
 * const auth = new OAuthStrategy({ clientId: '...', clientSecret: '...' });
 * const client = createMetricsClient({ shortCode: 'kv7kzm78', tenantId: 'zzxy_prd' }, auth);
 *
 * const { data, error } = await client.GET('/organizations/{organizationId}/metrics/overall', {
 *   params: { path: { organizationId: toOrganizationId('zzxy_prd') } },
 * });
 * ```
 *
 * @see {@link createMetricsClient} for instantiation
 */
export type MetricsClient = Client<paths>;

/**
 * Helper type to extract response data from an operation.
 */
export type MetricsResponse<T> = T extends {content: {'application/json': infer R}} ? R : never;

/**
 * The metrics data response envelope returned by every metrics operation.
 */
export type MetricsDataResponse = components['schemas']['MetricsDataResponse'];

/**
 * A single metric with its metadata and one or more data series.
 */
export type Metric = components['schemas']['Metric'];

/**
 * A named series of data points belonging to a {@link Metric}.
 */
export type MetricDataSeries = components['schemas']['DataSeries'];

/**
 * A single `{timestamp, value}` data point. `timestamp` is epoch milliseconds.
 */
export type MetricDataPoint = components['schemas']['DataPoint'];

/**
 * Standard RFC 7807 problem-details error response from the Metrics API.
 */
export type MetricsError = components['schemas']['ErrorResponse'];

/** Default OAuth scope required for the Metrics API (read-only). */
export const METRICS_DEFAULT_SCOPES = ['sfcc.metrics'];

/**
 * Configuration for creating a Metrics API client.
 */
export interface MetricsClientConfig {
  /**
   * The short code for the SCAPI instance.
   * This is typically a 4-8 character alphanumeric code.
   * @example "kv7kzm78"
   */
  shortCode: string;

  /**
   * The tenant ID (with or without f_ecom_ prefix).
   * Used to build the organizationId path parameter and tenant-specific OAuth scope.
   * @example "zzxy_prd" or "f_ecom_zzxy_prd"
   */
  tenantId: string;

  /**
   * Optional scope override. If not provided, defaults to the domain scope
   * (sfcc.metrics) plus the tenant-specific scope (SALESFORCE_COMMERCE_API:{tenant}).
   */
  scopes?: string[];

  /**
   * Middleware registry to use for this client.
   * If not specified, uses the global middleware registry.
   */
  middlewareRegistry?: MiddlewareRegistry;
}

/**
 * Creates a typed Metrics API client.
 *
 * **CLOSED BETA** — see the module description.
 *
 * Returns the openapi-fetch client directly, with authentication handled via
 * middleware. This gives full access to all openapi-fetch features with
 * type-safe paths, parameters, and responses.
 *
 * The client automatically handles OAuth scope requirements:
 * - Domain scope: `sfcc.metrics` (or custom via config.scopes)
 * - Tenant scope: `SALESFORCE_COMMERCE_API:{tenantId}`
 *
 * @param config - Metrics client configuration including shortCode and tenantId
 * @param auth - Authentication strategy (typically OAuth client-credentials)
 * @returns Typed openapi-fetch client
 *
 * @example
 * const auth = new OAuthStrategy({ clientId: '...', clientSecret: '...' });
 * const client = createMetricsClient({ shortCode: 'kv7kzm78', tenantId: 'zzxy_prd' }, auth);
 */
export function createMetricsClient(config: MetricsClientConfig, auth: AuthStrategy): MetricsClient {
  const registry = config.middlewareRegistry ?? globalMiddlewareRegistry;

  const client = createClient<paths>({
    baseUrl: `https://${config.shortCode}.api.commercecloud.salesforce.com/observability/metrics/v1`,
  });

  // Build required scopes: domain scope + tenant-specific scope
  const requiredScopes = config.scopes ?? [...METRICS_DEFAULT_SCOPES, buildTenantScope(config.tenantId)];

  // If auth supports scopes, add required scopes; otherwise use as-is
  const scopedAuth =
    auth instanceof OAuthStrategy || auth instanceof JwtOAuthStrategy
      ? auth.withAdditionalScopes(requiredScopes)
      : auth;

  // Core middleware: auth first
  client.use(createAuthMiddleware(scopedAuth));

  // Plugin middleware from registry
  for (const middleware of registry.getMiddleware('metrics')) {
    client.use(middleware);
  }

  // Logging middleware last (sees complete request with all modifications)
  client.use(createLoggingMiddleware('METRICS'));

  return client;
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/clients/metrics.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Metrics API client for B2C Commerce (Observability).

Mirrors ``src/clients/metrics.ts``. Provides a client for the SCAPI
Observability Metrics API (``observability/metrics/v1``). The API exposes
time-series operational metrics (overall, sales, eCDN, third-party services,
SCAPI, SCAPI hooks, MRT, controllers, and OCAPI) for an organization.

Only the client factory is ported here; the metrics seconds/ms window
resolution operations logic (if any) lives elsewhere and is out of scope.

Returns a configured :class:`~b2c_tooling_sdk.clients._core.HttpClient`; callers
use ``client.get(path, {...})`` etc., each returning a
:class:`~b2c_tooling_sdk.clients._core.ClientResult`.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from b2c_tooling_sdk.auth.types import AuthStrategy
from b2c_tooling_sdk.clients._core import HttpClient
from b2c_tooling_sdk.clients.custom_apis import build_tenant_scope
from b2c_tooling_sdk.clients.middleware import create_auth_middleware, create_logging_middleware
from b2c_tooling_sdk.clients.middleware_registry import MiddlewareRegistry, global_middleware_registry
from b2c_tooling_sdk.clients.models.metrics import (
    DataPoint as MetricDataPoint,
)
from b2c_tooling_sdk.clients.models.metrics import (
    DataSeries as MetricDataSeries,
)
from b2c_tooling_sdk.clients.models.metrics import (
    ErrorResponse as MetricsError,
)
from b2c_tooling_sdk.clients.models.metrics import (
    Metric,
    MetricsDataResponse,
)
from b2c_tooling_sdk.clients.scapi_backend_utils import with_scopes

#: The typed Metrics client. Aliased to :class:`HttpClient` (the
#: ``openapi-fetch`` ``Client`` analog).
MetricsClient = HttpClient

#: Default OAuth scope required for the Metrics API (read-only).
METRICS_DEFAULT_SCOPES = ["sfcc.metrics"]


@dataclass
class MetricsClientConfig:
    """Configuration for creating a Metrics API client."""

    #: The short code for the SCAPI instance (typically 4-8 alphanumeric chars).
    short_code: str
    #: The tenant ID (with or without ``f_ecom_`` prefix). Used to build the
    #: ``organizationId`` path parameter and tenant-specific OAuth scope.
    tenant_id: str
    #: Optional scope override. Defaults to the domain scope plus tenant scope.
    scopes: list[str] | None = None
    #: Optional middleware registry override (mainly for tests).
    middleware_registry: MiddlewareRegistry | None = field(default=None)


def create_metrics_client(config: MetricsClientConfig, auth: AuthStrategy) -> MetricsClient:
    """Create a typed Metrics API client.

    The client automatically handles OAuth scope requirements:

    - Domain scope: ``sfcc.metrics`` (or custom via ``config.scopes``)
    - Tenant scope: ``SALESFORCE_COMMERCE_API:{tenant_id}``

    :param config: Metrics client configuration including short code and tenant ID.
    :param auth: Authentication strategy (typically OAuth client-credentials).
    :returns: A configured :class:`HttpClient`.
    """
    registry = config.middleware_registry or global_middleware_registry

    client = HttpClient(
        f"https://{config.short_code}.api.commercecloud.salesforce.com/observability/metrics/v1",
        client_type="metrics",
    )

    # Build required scopes: domain scope + tenant-specific scope.
    required_scopes = config.scopes or [*METRICS_DEFAULT_SCOPES, build_tenant_scope(config.tenant_id)]

    # If auth supports scopes, add required scopes; otherwise use as-is.
    scoped_auth = with_scopes(auth, required_scopes)

    # Core middleware: auth first.
    client.use(create_auth_middleware(scoped_auth))

    # Plugin middleware from the registry.
    for middleware in registry.get_middleware("metrics"):
        client.use(middleware)

    # Logging middleware last (sees the complete request with all modifications).
    client.use(create_logging_middleware("METRICS"))

    return client


__all__ = [
    "METRICS_DEFAULT_SCOPES",
    "Metric",
    "MetricDataPoint",
    "MetricDataSeries",
    "MetricsClient",
    "MetricsClientConfig",
    "MetricsDataResponse",
    "MetricsError",
    "create_metrics_client",
]
